akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala

```diff
@@ -0,0 +1,130 @@
+  implicit val materializer = ActorMaterializer(settings)
+
+  val timeout = 300.milliseconds
+  val bytes = "abc".getBytes
+  val byteString = ByteString("abc")
+
+  def expectBlocked[T](f: Future[T]) =
+    the[Exception] thrownBy Await.result(f, timeout) shouldBe a[TimeoutException]
+
+  def expectSuccess[T](f: Future[T], value: T) =
+    Await.result(f, timeout) should be(value)
+
+  "OutputStreamSource" must {
+    "read bytes from OutputStream" in assertAllStagesStopped {
+      val ((outputStream, f), probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+      s.request(1)
+      probe.expectNext(byteString)
+      outputStream.close()
+      probe.expectComplete()
+      Await.result(f, timeout) should be(3)
+    }
+
+    "block flush call until send all buffer to downstream" in assertAllStagesStopped {
+      val ((outputStream, _), probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+      val f = Future(outputStream.flush())
+
+      probe.expectNoMsg(timeout)
+      expectBlocked(f)
+
+      s.request(1)
+      expectSuccess(f, ())
+      probe.expectNext(byteString)
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "not block flushes when buffer is empty" in assertAllStagesStopped {
+      val ((outputStream, _), probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+
+      val f = Future(outputStream.flush())
+      s.request(1)
+      expectSuccess(f, ())
+      probe.expectNext(byteString)
+
+      val f2 = Future(outputStream.flush())
+      expectSuccess(f2, ())
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "block writes when buffer is full" in assertAllStagesStopped {
+      val ((outputStream, _), probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      //default max-input-buffer-size = 16
+      (1 to 15).foreach { _ ⇒ outputStream.write(bytes) }
+
+      //blocked call
+      val f = Future(outputStream.write(bytes))
+
+      probe.expectNoMsg(timeout)
+      expectBlocked(f)
+
+      s.request(16)
+      expectSuccess(f, ())
+      probe.expectNextN(List.fill(16)(byteString).toSeq)
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "throw error when reactive stream is closed" in assertAllStagesStopped {
+      val ((outputStream, _), probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+
+      probe.expectSubscription()
+      outputStream.close()
+      probe.expectComplete()
+      //source of dead letters
+      the[Exception] thrownBy outputStream.write(bytes) shouldBe a[IOException]
+    }
```

Comment:
there’s a test case missing: what about writing to the stream after the subscriber has cancelled the reactive stream?